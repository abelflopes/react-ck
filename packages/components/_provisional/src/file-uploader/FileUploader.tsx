import styles from "./styles/index.module.scss";
import React, { useMemo, useRef } from "react";
import classNames from "classnames";
import { Text } from "@react-ck/text";
import { Button, type ButtonProps } from "@react-ck/button";
import { readFileList } from "./utils/read-file";

// TODO: check https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/file
// TODO: add size limitation: https://stackoverflow.com/questions/5697605/limit-the-size-of-a-file-upload-html-input-element
// TODO: max number of files
// TODO: size display helper
// TODO: util convert url to file https://stackoverflow.com/questions/35940290/how-to-convert-base64-string-to-javascript-file-object-like-as-from-file-input-f
// TODO: util file size description
// TODO: add overlay loader
// TODO: add uploaded files feedback

export interface FileUploaderProps extends Omit<React.HTMLAttributes<HTMLElement>, "onChange"> {
  skin?: "default" | "negative" | "disabled";
  variation?: "default" | "square";
  icon?: React.ReactNode;
  cta?: React.ReactNode;
  description?: React.ReactNode;
  inputProps?: Omit<React.InputHTMLAttributes<HTMLInputElement>, "type">;
  buttonProps?: ButtonProps;
  /** The validation message text */
  validationMessage?: React.ReactNode;
  onChange: (
    e: React.ChangeEvent<HTMLInputElement>,
    fileList: ReturnType<typeof readFileList>,
  ) => void;
}

// eslint-disable-next-line complexity -- TODO: fix
export const FileUploader = ({
  skin = "default",
  variation = "default",
  icon,
  cta,
  description,
  validationMessage,
  className,
  children,
  inputProps,
  buttonProps,
  onChange,
  ...otherProps
}: Readonly<FileUploaderProps>): React.ReactElement => {
  const inputRef = useRef<HTMLInputElement>(null);

  const isIconOnly = useMemo(() => Boolean(icon) && !cta && !description, [cta, description, icon]);

  const onEnterPress = (e: React.KeyboardEvent<HTMLElement>): void => {
    if (e.code === "Enter") inputRef.current?.click();
  };

  return (
    <div
      {...otherProps}
      className={classNames(
        variation !== "default" && styles[variation],
        skin !== "default" && styles[skin],
        !isIconOnly && styles.root,
        isIconOnly && styles.root_icon_only,
        className,
      )}>
      <input
        {...inputProps}
        ref={inputRef}
        type="file"
        className={classNames(styles.file, inputProps?.className)}
        onKeyUp={onEnterPress}
        onChange={(e) => {
          if (!e.target.files) throw new Error("No files added");
          onChange(e, readFileList(e.target.files));
        }}
      />
      {!isIconOnly && icon}
      {children ? <div className={styles.content}>{children}</div> : null}
      {isIconOnly || cta ? (
        <Button
          {...buttonProps}
          icon={isIconOnly ? icon : null}
          disabled={skin === "disabled" || buttonProps?.disabled}
          className={classNames(styles.button, buttonProps?.className)}
          onKeyUp={(e) => {
            onEnterPress(e);
            buttonProps?.onKeyUp?.(e);
          }}
          onClick={(e) => {
            inputRef.current?.click();
            buttonProps?.onClick?.(e);
          }}>
          {cta}
        </Button>
      ) : null}

      {description || validationMessage ? (
        <div className={styles.details}>
          {description ? <Text variation="small">{description}</Text> : null}
          {validationMessage ? (
            <Text variation="small" className={styles.validation_message}>
              {validationMessage}
            </Text>
          ) : null}
        </div>
      ) : null}
    </div>
  );
};
