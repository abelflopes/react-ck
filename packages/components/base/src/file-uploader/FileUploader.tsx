import styles from "./styles/index.module.scss";
import React, { useCallback, useMemo, useRef, useEffect } from "react";
import classNames from "classnames";
import { Text } from "../text";
import { Button, type ButtonProps } from "../button";
import { readFileList } from "./utils/read-file";
import { megeRefs } from "@react-ck/react-utils";

// TODO: check https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/file
// TODO: add size limitation: https://stackoverflow.com/questions/5697605/limit-the-size-of-a-file-upload-html-input-element
// TODO: max number of files
// TODO: size display helper
// TODO: util convert url to file https://stackoverflow.com/questions/35940290/how-to-convert-base64-string-to-javascript-file-object-like-as-from-file-input-f
// TODO: util file size description
// TODO: add overlay loader
// TODO: add uploaded files feedback

/**
 * Props for configuring the FileUploader component
 */
interface FileUploaderProps
  extends Omit<React.ComponentPropsWithRef<"div">, "onChange" | "onProgress" | "type"> {
  /** Visual style of the uploader. Defaults to "default" */
  skin?: "default" | "negative" | "disabled";
  /** Layout variation of the uploader. Defaults to "default" */
  variation?: "default" | "square";
  /** Optional icon to display */
  icon?: React.ReactNode;
  /** Optional description text */
  description?: React.ReactNode;
  /** Props passed to the file input element */
  inputProps?: Omit<React.ComponentPropsWithRef<"input">, "type">;
  /** Props passed to the button element when in icon-only mode */
  buttonProps?: ButtonProps;
  /** Validation message displayed below the uploader */
  validationMessage?: React.ReactNode;
  /** Callback fired when files are selected */
  onChange?: (
    e: React.ChangeEvent<HTMLInputElement>,
    fileList: ReturnType<typeof readFileList>,
  ) => void;
  /** Callback fired during file reading progress */
  onProgress?: Parameters<typeof readFileList>[1];
  /** Try topen the file input immediately when the component is mounted */
  openImmediately?: boolean;
}

/**
 * File input component with drag and drop support
 * Supports icon-only and descriptive layouts with validation
 */
export const FileUploader = ({
  skin = "default",
  variation = "default",
  icon,
  description,
  validationMessage,
  className,
  children,
  inputProps,
  buttonProps,
  onChange,
  onProgress,
  openImmediately,
  ...otherProps
}: Readonly<FileUploaderProps>): React.ReactElement => {
  const inputRef = useRef<HTMLInputElement>(null);
  const isIconOnly = useMemo(() => Boolean(icon) && !description, [description, icon]);

  const isFirstRender = useRef(true);

  const handleKeyUp = useCallback((e: React.KeyboardEvent<HTMLElement>): void => {
    if (e.code === "Enter") inputRef.current?.click();
  }, []);

  const handleChange = useCallback<React.ChangeEventHandler<HTMLInputElement>>(
    (e) => {
      if (!e.target.files) throw new Error("No files added");

      const fileList = readFileList(e.target.files, onProgress);

      onChange?.(e, fileList);

      inputProps?.onChange?.(e);
    },
    [inputProps, onChange, onProgress],
  );

  useEffect(() => {
    if (openImmediately && isFirstRender.current) inputRef.current?.click();
  }, [openImmediately]);

  useEffect(() => {
    isFirstRender.current = false;
  }, []);

  return (
    <div
      {...otherProps}
      className={classNames(
        variation !== "default" && styles[variation],
        skin !== "default" && styles[skin],
        !isIconOnly && styles.root,
        isIconOnly && styles.root_icon_only,
        className,
      )}>
      <input
        {...inputProps}
        ref={megeRefs(inputRef, inputProps?.ref)}
        type="file"
        className={classNames(styles.file, inputProps?.className)}
        onChange={handleChange}
        onKeyUp={(e) => {
          handleKeyUp(e);
          inputProps?.onKeyUp?.(e);
        }}
      />

      {!isIconOnly && icon}

      {children ? <div className={styles.content}>{children}</div> : null}

      {isIconOnly ? (
        <Button
          {...buttonProps}
          icon={icon}
          disabled={skin === "disabled" || buttonProps?.disabled}
          className={classNames(styles.button, buttonProps?.className)}
          onKeyUp={(e) => {
            handleKeyUp(e);
            buttonProps?.onKeyUp?.(e);
          }}
          onClick={(e) => {
            inputRef.current?.click();
            buttonProps?.onClick?.(e);
          }}
        />
      ) : null}

      {description || validationMessage ? (
        <div className={styles.details}>
          {description ? (
            <Text variation="small" skin="soft">
              {description}
            </Text>
          ) : null}
          {validationMessage ? (
            <Text variation="small" className={styles.validation_message}>
              {validationMessage}
            </Text>
          ) : null}
        </div>
      ) : null}
    </div>
  );
};

export { type FileUploaderProps };
