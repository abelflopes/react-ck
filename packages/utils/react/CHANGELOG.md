# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [1.6.0](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.5.0...@react-ck/react-utils@1.6.0) (2025-07-09)


### Features

* add expandable component ([e5cd8e7](https://github.com/abelflopes/react-ck/commit/e5cd8e7c3362eb80bf798ffa83757b35d9b43df2))



## [1.5.0](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.4.0...@react-ck/react-utils@1.5.0) (2025-06-07)


### Features

* support focus trapping and apply consistent focus outline ([26ea216](https://github.com/abelflopes/react-ck/commit/26ea2167102a4d1c89f9bd937dc44949ef2a8b0e))



## [1.4.0](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.3.7...@react-ck/react-utils@1.4.0) (2024-12-19)


### Features

* **react utils:** add ref merge util ([42683b3](https://github.com/abelflopes/react-ck/commit/42683b3cf99617076eb29588fd6b6d94bf842b2f))



## [1.3.7](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.3.6...@react-ck/react-utils@1.3.7) (2024-10-09)


### Bug Fixes

* update react version ([59e4e9a](https://github.com/abelflopes/react-ck/commit/59e4e9afa979d29efdc793f3441ed528971844ca))



## [1.3.6](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.3.5...@react-ck/react-utils@1.3.6) (2024-07-10)

**Note:** Version bump only for package @react-ck/react-utils





## [1.3.5](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.3.4...@react-ck/react-utils@1.3.5) (2024-07-10)

**Note:** Version bump only for package @react-ck/react-utils





## [1.3.4](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.3.3...@react-ck/react-utils@1.3.4) (2024-07-09)

**Note:** Version bump only for package @react-ck/react-utils





## [1.3.3](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.3.2...@react-ck/react-utils@1.3.3) (2024-07-08)

**Note:** Version bump only for package @react-ck/react-utils





## [1.3.2](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.3.1...@react-ck/react-utils@1.3.2) (2024-07-05)

**Note:** Version bump only for package @react-ck/react-utils





## [1.3.1](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.3.0...@react-ck/react-utils@1.3.1) (2024-07-05)


### Bug Fixes

* component identification with display name ([b182afb](https://github.com/abelflopes/react-ck/commit/b182afb3468d0ca00d9923f05d678bc8db8492f0))



## [1.3.0](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.2.12...@react-ck/react-utils@1.3.0) (2024-07-04)


### Features

* add click outside util ([738a2aa](https://github.com/abelflopes/react-ck/commit/738a2aab13db43fbe4a64f41030ec40fd6fa7074))
* add next render util ([30ca88a](https://github.com/abelflopes/react-ck/commit/30ca88a03ec1c760097331a4f48857b0d7b017d2))
* update react utils ([75c4350](https://github.com/abelflopes/react-ck/commit/75c43505ddba5382cc71e619588764abf179d84f))



## [1.2.12](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.2.11...@react-ck/react-utils@1.2.12) (2024-06-26)

**Note:** Version bump only for package @react-ck/react-utils





## [1.2.11](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.2.10...@react-ck/react-utils@1.2.11) (2024-06-23)

**Note:** Version bump only for package @react-ck/react-utils





## [1.2.10](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.2.9...@react-ck/react-utils@1.2.10) (2024-06-15)

**Note:** Version bump only for package @react-ck/react-utils





## [1.2.9](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.2.8...@react-ck/react-utils@1.2.9) (2024-03-19)

**Note:** Version bump only for package @react-ck/react-utils





## [1.2.8](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.2.7...@react-ck/react-utils@1.2.8) (2024-03-17)

**Note:** Version bump only for package @react-ck/react-utils





## [1.2.7](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.2.6...@react-ck/react-utils@1.2.7) (2024-03-17)

**Note:** Version bump only for package @react-ck/react-utils





## [1.2.6](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.2.5...@react-ck/react-utils@1.2.6) (2024-03-16)

**Note:** Version bump only for package @react-ck/react-utils





## [1.2.5](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.2.4...@react-ck/react-utils@1.2.5) (2024-03-15)

**Note:** Version bump only for package @react-ck/react-utils





## [1.2.4](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.2.3...@react-ck/react-utils@1.2.4) (2024-03-15)

**Note:** Version bump only for package @react-ck/react-utils





## [1.2.3](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.2.2...@react-ck/react-utils@1.2.3) (2024-03-15)

**Note:** Version bump only for package @react-ck/react-utils





## [1.2.2](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.2.1...@react-ck/react-utils@1.2.2) (2024-03-12)

**Note:** Version bump only for package @react-ck/react-utils





## [1.2.1](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.2.0...@react-ck/react-utils@1.2.1) (2024-03-12)

**Note:** Version bump only for package @react-ck/react-utils





## [1.2.0](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.1.1...@react-ck/react-utils@1.2.0) (2024-03-06)


### Features

* add component polimorphy util ([e653799](https://github.com/abelflopes/react-ck/commit/e65379984d022ca716a3b040aae03aa65528fcaa))
* add util to convert component into text ([4580f6c](https://github.com/abelflopes/react-ck/commit/4580f6c4630d179b2dda084cfaa04c518f551efc))


### Bug Fixes

* child list misssing key warning ([e05e301](https://github.com/abelflopes/react-ck/commit/e05e3019493b01bf492aa13de775adc7891faa35))



## [1.1.1](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.1.0...@react-ck/react-utils@1.1.1) (2024-01-24)

**Note:** Version bump only for package @react-ck/react-utils





## 1.1.0 (2024-01-02)


### Features

* add react utils package ([fe8737f](https://github.com/abelflopes/react-ck/commit/fe8737f2bf784a0b42111a03980112fa3d4f0795))


### Bug Fixes

* make packages public ([16c94d3](https://github.com/abelflopes/react-ck/commit/16c94d3bd288a46084896f7fa6f2be33a3609e4f))
