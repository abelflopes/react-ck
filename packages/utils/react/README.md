# React CK | Utils Library

> :warning: **WARNING**: This component library is being updated frequently and it's currently unstable due to being in it's early stages, we advice you to use only in production environments only after version **2.0.0**.

### Installation 

To integrate the this library into your React apps, you can install it using npm or yarn: `npm i --save @react-ck/react-utils` or `yarn add @react-ck/react-utils`.

## Utilities

### Polimorphy

Component polimorphy allows consumer to choose which HTML tags and attributes will be rendered from a specific component.

### Usage Examples

### Implementation

```diff
import React, { isValidElement, useEffect, useMemo } from "react";
import styles from "./styles/index.module.scss";
import classNames from "classnames";
- import { getDisplayName } from "@react-ck/react-utils";
+ import {
+  getDisplayName,
+  PolymorphicComponent,
+  type ConsumerPolymorphicProps,
+  type HTMLTag,
+ } from "@react-ck/react-utils";

- export interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
+ export interface ButtonProps<T extends HTMLTag>
+  extends React.ButtonHTMLAttributes<HTMLButtonElement>,
+    ConsumerPolymorphicProps<T> {
  skin?: "primary" | "secondary" | "ghost" | "negative";
  size?: "s" | "m" | "l";
  icon?: React.ReactNode;
}

- export const Button = ({
+ export const Button = <T extends HTMLTag>({
+  as,
  skin = "primary",
  size = "m",
  icon,
  children,
  className,
  ...otherProps
- }: Readonly<ButtonProps>): React.ReactElement => {
+ }: Readonly<ButtonProps<T>>): React.ReactElement => {
  const isIconOnly = useMemo(
    () => icon !== undefined && React.Children.count(children) === 0,
    [children, icon],
  );

  return (
-    <button
-      className={classNames(
-        styles.root,
-        styles[skin],
-        styles[`size-${size}`],
-        {
-          [`${styles["icon-only"]}`]: isIconOnly,
-        },
-        className,
-      )}
-      {...otherProps}>
+    <PolymorphicComponent
+      as={as}
+      fallback={["button", otherProps]}
+      commonProps={{
+        className: classNames(
+          styles.root,
+          styles[skin],
+          styles[`size-${size}`],
+          {
+            [`${styles["icon-only"]}`]: isIconOnly,
+          },
+          className,
+        ),
+      }}>
      {icon && !isIconOnly ? <span className={styles.icon}>{icon}</span> : null}
      {children}
      {isIconOnly ? icon : null}
-     </button>
+    </PolymorphicComponent>
  );
};
```


<!-- storybook-ignore -->

---

Check the documentation website - [react-ck.js.org](https://react-ck.js.org).