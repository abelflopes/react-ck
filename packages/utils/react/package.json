{
  "name": "@react-ck/react-utils",
  "private": false,
  "version": "1.3.6",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "files": [
    "/dist"
  ],
  "homepage": "https://github.com/abelflopes/react-ck/tree/master/packages/utils/react-hooks#readme",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/abelflopes/react-ck.git"
  },
  "scripts": {
    "build": "tsc -p tsconfig.build.json",
    "lint:typescript": "tsc --noEmit",
    "lint:code": "eslint . --ext ts,tsx --cache",
    "test": "jest --testPathPattern=\".unit.*\""
  },
  "devDependencies": {
    "@react-ck/babel-config": "^1.0.0",
    "@react-ck/jest-config": "^1.0.0",
    "@react-ck/typescript-config": "^1.0.0",
    "@types/react": "^18.3.11"
  },
  "peerDependencies": {
    "react": "^18.3.1"
  },
  "dependencies": {
    "classnames": "^2.5.1"
  }
}
