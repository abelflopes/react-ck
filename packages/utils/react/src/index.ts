export * from "./display-name";

export * from "./polimorphy";

export * from "./component-to-text";

export * from "./click-outside";

export * from "./children-without-fragments";

export * from "./merge-refs";

export * from "./raf";

export * from "./promise";
