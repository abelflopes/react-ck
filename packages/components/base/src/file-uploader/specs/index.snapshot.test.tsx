import React from "react";
import { FileUploader } from "../FileUploader";
import { render } from "@testing-library/react";

describe("snapshot FileUploader", () => {
  it("should match snapshot", () => {
    const { container } = render(<FileUploader />);
    expect(container).toMatchSnapshot();
  });
});
