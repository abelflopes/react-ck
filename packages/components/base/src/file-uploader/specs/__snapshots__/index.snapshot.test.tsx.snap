// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`snapshot FileUploader should match snapshot 1`] = `
<div>
  <div
    class="root"
  >
    <input
      class="file"
      type="file"
    />
  </div>
</div>
`;
